### PR TITLE
Add patches to make LLVM use spack zlib/ncurses

### LINKED ISSUE #17981: LLVM's support lib cmake targets have dependencies on system libs (zlib, ncurses/tinfo)
When building `spack install llvm`, it creates a cmake file:

https://github.com/llvm/llvm-project/blob/master/llvm/lib/Support/CMakeLists.txt#L1

which is installed to e.g.

```
opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/llvm-10.0.0-576xf5b5mlqxsskl45ucl3xxxmjxssio/lib/cmake/llvm
```
It contains certain strict dependencies on 'system libraries':

```
$ grep tinfo LLVMExports.cmake 
  INTERFACE_LINK_LIBRARIES "z;rt;dl;tinfo;-lpthread;m;/home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/z3-4.8.7-qiwh5wxz2mdknzv4gef7lwxxdnlf5ycx/lib/libz3.so;LLVMDemangle"
```
only `z3` has a proper path to the spack installed lib.

In my case I was building rocBLAS, which uses this LLVM target, links against the system version of `libtinfo.so`, but gets an RPATH to the spack-version of it, and therefore it fails at runtime.

### PR DESCRIPTION
**Pull request by haampie**
This should fix https://github.com/spack/spack/issues/17981 for LLVM 6 and above, e.g.:

```
harmen-desktop spack-stage $ grep "libz" ./spack-stage-llvm-10.0.0-kmr2veey2khsmolq5izpz2jsrb2oya5b/spack-build/lib/cmake/llvm/LLVMExports.cmake 
  INTERFACE_LINK_LIBRARIES "/home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/zlib-1.2.11-2pwsgfxppopolmjj6tf34k5jsaqzpodo/lib/libz.so;rt;dl;/home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/ncurses-6.2-l4seuemvhefwetlot2dbcnlaxtncqzqd/lib/libtinfo.so;-lpthread;m;/home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/z3-4.8.7-qiwh5wxz2mdknzv4gef7lwxxdnlf5ycx/lib/libz3.so;LLVMDemangle"
harmen-desktop spack-stage $ grep "libz" ./spack-stage-llvm-7.0.1-u4vs4lbw3bt664lzzm66rigz2detyduh/spack-build/lib/cmake/llvm/LLVMExports.cmake
  INTERFACE_LINK_LIBRARIES "/home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/zlib-1.2.11-2pwsgfxppopolmjj6tf34k5jsaqzpodo/lib/libz.so;rt;dl;/home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/ncurses-6.2-l4seuemvhefwetlot2dbcnlaxtncqzqd/lib/libtinfo.so;-lpthread;m;LLVMDemangle"

```

I'll try to submit the same patch upstream as well

### CODE REVIEW

**Comment by adamjstewart** [comment]
@trws @naromero77 can you review this patch?

**Comment by naromero77** [comment]
I am OK with either using this patch, as long as it is functionally equivalent to these 
https://reviews.llvm.org/D79219
https://reviews.llvm.org/D85820

@trws any additional concerns?

**Comment by haampie** [comment]
@trws response was here I think: https://github.com/spack/spack/issues/17981#issuecomment-673048625

**Comment by trws** [comment]
Not immediately, I think we'll want to limit this patch to versions that lack those LLVM patches once they go live.  So maybe have a tracking issue for followup or something, but I think the change is positive.

**Comment by haampie** [comment]
Let me at the very least add a comment in the `package.py` with links to  https://reviews.llvm.org/D79219 and  https://reviews.llvm.org/D85820

**Comment by haampie** [comment]
@trws @naromero77: I would want to close this PR. My changes have landed in LLVM's master branch now, but it was much more involved to get it right. It looks like a trivial change, but in fact it took 3 or 4 iterations to make sure LLVM's OCaml bindings would still build, and there were issues with Chrome on Windows and what not :laughing:. That's all fixed now, and we could actually backport those specific changes, but the problem is we would really have to do the same for `zlib`, and there's been quite some refactoring in how zlib is handled.

So TL;DR: it looks like a trivial patch, it really is not because of how big LLVM is, so rather just wait for a new release with battle-tested cmake scripts.

---

I believe clang 11 just branched off, not sure if my patches land in it, but probably it could be backported.

**Comment by haampie** [comment]
Unfortunately the patches for having non-system zlib and ncurses in LLVM will not be backported to LLVM 11, so they will only land in LLVM 12 :( ref https://reviews.llvm.org/rG31e5f7120bdd2f76337686d9d169b1c00e6ee69c#942683.

So, I'll try to just cherry-pick the commits related to zlib / ncurses that currently live on the LLVM master branch, and if they do not cause conflicts on LLVM 10, I'll commit them and reopen this PR.
